libnopegl/src/gen_specs.c

```diff
@@ -19,29 +19,156 @@
+                    printf(D "%d,\n", p->def_value.i32);
+                break;
+            case NGLI_PARAM_TYPE_I32:
+                printf(D "%d,\n", p->def_value.i32);
+                break;
+            case NGLI_PARAM_TYPE_U32:
+                printf(D "%u,\n", p->def_value.u32);
+                break;
+            case NGLI_PARAM_TYPE_IVEC2: printf(D "[%d,%d],\n",       NGLI_ARG_VEC2(p->def_value.ivec)); break;
+            case NGLI_PARAM_TYPE_IVEC3: printf(D "[%d,%d,%d],\n",    NGLI_ARG_VEC3(p->def_value.ivec)); break;
+            case NGLI_PARAM_TYPE_IVEC4: printf(D "[%d,%d,%d,%d],\n", NGLI_ARG_VEC4(p->def_value.ivec)); break;
+            case NGLI_PARAM_TYPE_UVEC2: printf(D "[%u,%u],\n",       NGLI_ARG_VEC2(p->def_value.uvec)); break;
+            case NGLI_PARAM_TYPE_UVEC3: printf(D "[%u,%u,%u],\n",    NGLI_ARG_VEC3(p->def_value.uvec)); break;
+            case NGLI_PARAM_TYPE_UVEC4: printf(D "[%u,%u,%u,%u],\n", NGLI_ARG_VEC4(p->def_value.uvec)); break;
+            case NGLI_PARAM_TYPE_VEC2:  printf(D "[%f,%f],\n",       NGLI_ARG_VEC2(p->def_value.vec));  break;
+            case NGLI_PARAM_TYPE_VEC3:  printf(D "[%f,%f,%f],\n",    NGLI_ARG_VEC3(p->def_value.vec));  break;
+            case NGLI_PARAM_TYPE_VEC4:  printf(D "[%f,%f,%f,%f],\n", NGLI_ARG_VEC4(p->def_value.vec));  break;
+            case NGLI_PARAM_TYPE_MAT4:
+                printf(D "[%f,%f,%f,%f,%f,%f,%f,%f,%f,%f,%f,%f,%f,%f,%f,%f],\n", NGLI_ARG_MAT4(p->def_value.mat));
+                break;
+            case NGLI_PARAM_TYPE_STR:
+                if (p->def_value.str)
+                    printf(D "\"%s\",\n", p->def_value.str);
+                break;
+            case NGLI_PARAM_TYPE_RATIONAL:
+                printf(D "[%d,%d],\n", NGLI_ARG_VEC2(p->def_value.r));
+                break;
+            }
+            if (p->node_types) {
+                printf("      \"node_types\": [");
+                for (int i = 0; p->node_types[i] != -1; i++) {
+                    const struct node_class *cls = get_node_class(p->node_types[i]);
+                    ngli_assert(cls);
+                    printf("%s\"%s\"", i ? ", " : "", cls->name);
+                }
+                printf("],\n");
+            }
+            static const struct {
+                const char *name;
+                const int flag;
+            } flag_names[] = {
+                {"live", NGLI_PARAM_FLAG_ALLOW_LIVE_CHANGE},
+                {"node", NGLI_PARAM_FLAG_ALLOW_NODE},
+                {"nonull", NGLI_PARAM_FLAG_NON_NULL},
+            };
+            printf("      \"flags\": [");
+            int ihasflag = 0;
+            for (int i = 0; i < NGLI_ARRAY_NB(flag_names); i++) {
+                if (!(p->flags & flag_names[i].flag))
+                    continue;
+                printf("%s\"%s\"", ihasflag ? ", " : "", flag_names[i].name);
+                ihasflag = 1;
+            }
+            printf("],\n");
+            char *esc_desc = escape_json_str(p->desc);
```

Comment:
D'ont we want to error out in case of alloc failure instead of generating an incomplete spec ?

Reply:
`escape_json_str()` accepts a `NULL` string and makes no distinction with a malloc failure. Since this is a dummy tool I took the liberty to add an `abort()` in case on malloc failure in this function. I hope that's acceptable. I did the same thing in the function itself here:
```c
            case NGLI_PARAM_TYPE_FLAGS: {
                const int v = (int)p->def_value.i32;
                char *s = ngli_params_get_flags_str(p->choices->consts, v);
                if (!s)
                    abort();
```

Alternatively I could add a more advanced error management, but that would imply adding a lot of checks for limited benefit since this is not a program relying on user input.

Reply:
That's fine as is. Thanks.